File: flutter/lib/domain/usecase/list_products_use_case.dart
// flutter/lib/domain/usecase/list_products_use_case.dart
import 'package:grpc_demo/domain/model/product.dart';
import 'package:grpc_demo/domain/repository/catalog_repository.dart';

class ListProductsUseCase {
  const ListProductsUseCase(this._repository);
  final CatalogRepository _repository;

  Stream<Product> call({String category = '', int page = 1}) =>
      _repository.listProducts(category: category, page: page);
}

File: android/app/src/main/kotlin/com/demo/grpcdemo/domain/usecase/ListProductsUseCase.kt
package com.demo.grpcdemo.domain.usecase

import com.demo.grpcdemo.domain.model.Product
import com.demo.grpcdemo.domain.repository.CatalogRepository
import kotlinx.coroutines.flow.Flow
import javax.inject.Inject

class ListProductsUseCase @Inject constructor(private val repository: CatalogRepository) {
    operator fun invoke(category: String = "", page: Int = 1): Flow<Product> =
        repository.listProducts(category, page)
}
